File: src/main/java/me/sul/crackshotaddition/weapons/FlameThrower.java
package me.sul.crackshotaddition.weapons;

import com.shampaggon.crackshot.events.WeaponHitBlockEvent;
import com.shampaggon.crackshot.events.WeaponShootEvent;
import me.sul.crackshotaddition.CrackShotAddition;
import me.sul.crackshotaddition.events.CrackShotProjectileBlockBreakEffectEvent;
import me.sul.crackshotaddition.events.CrackShotProjectileTrailEvent;
import net.minecraft.server.v1_12_R1.PacketPlayOutWorldParticles;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.Particle;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.entity.Entity;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitRunnable;

public class FlameThrower implements Listener {
    public static final String FLAMETHROWER = "N화염방사기";

    @EventHandler
    public void onBlockBreakEffectEvent(CrackShotProjectileBlockBreakEffectEvent e) {
        if (e.getWeaponTitle().equals(FLAMETHROWER)) {
            e.setCancelled(true);
        }
    }

    @EventHandler
    public void onProjectileTrailEvent(CrackShotProjectileTrailEvent e) {
        if (e.getWeaponTitle().equals(FLAMETHROWER)) {
            e.setParticle(Particle.FLAME);
        }
    }

    @EventHandler
    public void onShoot(WeaponShootEvent e) {
        if (e.getWeaponTitle().equals(FLAMETHROWER)) {
            Entity proj = e.getProjectile();

            new BukkitRunnable() {
                PacketPlayOutWorldParticles packet;
                Location loc;
                Location previousLoc;
                boolean skipFirstLoc = true;
                int passedTick = 0;

                @Override
                public void run() {
                    if (proj.isValid() == false) cancel();
                    loc = proj.getLocation();
                    if (!skipFirstLoc) {
                        loc.getWorld().spawnParticle(Particle.LAVA, loc,1);
                    } else {
                        skipFirstLoc = false;
                    }

                    if (loc.equals(previousLoc)) {
                        cancel();
                    }
                    if (passedTick >= 10) {
                        e.getProjectile().remove();
                        cancel();
                    }

                    passedTick += 2;
                    previousLoc = loc;
                }
            }.runTaskTimer((Plugin) CrackShotAddition.getInstance(), 0, 2);
        }
    }

    @EventHandler
    public void onWeaponHitBlock(WeaponHitBlockEvent e) {
        if (e.getWeaponTitle().equals(FLAMETHROWER)) {
            Block block = e.getBlock();
            if (block.getRelative(BlockFace.UP).getType().equals(Material.AIR)) {
                block.getRelative(BlockFace.UP).setType(Material.FIRE);

                new BukkitRunnable() {
                    @Override
                    public void run() {
                        if (block.getRelative(BlockFace.UP).getType().equals(Material.FIRE)) {
                            block.getRelative(BlockFace.UP).setType(Material.AIR);
                        }
                    }
                }.runTaskLater(CrackShotAddition.getInstance(), 20);
            }
        }
    }
}

File: src/main/kotlin/me/sul/crackshotaddition/weapons/FlameThrower.kt
package me.sul.crackshotaddition.weapons

import com.shampaggon.crackshot.events.WeaponHitBlockEvent
import com.shampaggon.crackshot.events.WeaponShootEvent
import me.sul.crackshotaddition.CrackShotAddition
import net.minecraft.server.v1_12_R1.PacketPlayOutWorldParticles
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.block.BlockFace
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitRunnable

class FlameThrower : Listener {
    companion object {
        const val FLAMETHROWER = "N화염방사기"
    }

    @EventHandler
    fun onShoot(e: WeaponShootEvent) {
        if (e.weaponTitle == FLAMETHROWER) {
            val proj = e.projectile
            object : BukkitRunnable() {
                var previousLoc: Location? = null
                var skipFirstLoc = true
                var passedTick = 0
                override fun run() {
                    if (!proj.isValid) {
                        cancel(); return
                    }
                    val loc = proj.location
                    if (!skipFirstLoc) {
                        loc.getWorld().spawnParticle(Particle.LAVA, loc, 1)
                    } else {
                        skipFirstLoc = false
                    }
                    if (previousLoc != null && loc == previousLoc) {
                        cancel(); return
                    }
                    if (passedTick >= 10) {
                        e.projectile.remove()
                        cancel(); return
                    }
                    passedTick += 2
                    previousLoc = loc
                }
            }.runTaskTimer(CrackShotAddition.instance as Plugin, 0, 2)
        }
    }

    @EventHandler
    fun onWeaponHitBlock(e: WeaponHitBlockEvent) {
        if (e.weaponTitle == FLAMETHROWER) {
            val block = e.block
            if (block.getRelative(BlockFace.UP).type == Material.AIR) {
                block.getRelative(BlockFace.UP).type = Material.FIRE
                object : BukkitRunnable() {
                    override fun run() {
                        if (block.getRelative(BlockFace.UP).type == Material.FIRE) {
                            block.getRelative(BlockFace.UP).type = Material.AIR
                        }
                    }
                }.runTaskLater(CrackShotAddition.instance, 20)
            }
        }
    }
}